Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -15,709 +15,573 @@
+                                                                versionMinor: Int(cParser.http_minor),
+                                                                statusCode: Int(cParser.status_code),
+                                                                isUpgrade: cParser.upgrade != 0,
+                                                                method: http_method(rawValue: cParser.method)) {
+            case .normal:
+                return 0
+            case .skipBody:
+                return 1
+            case .error(let err):
+                parser.httpErrno = err
+                return -1 // error
+            }
+        }
+        self.settings.on_message_complete = { opaque in
+            BetterHTTPParser.fromOpaque(opaque).didReceiveMessageCompleteNotification()
+            return 0
         }
-
-        context.write(data, promise: promise)
-    }
-}
-
-/// Strategy to use when a HTTPDecoder is removed from a pipeline after a HTTP upgrade was detected.
-public enum RemoveAfterUpgradeStrategy {
-    /// Forward all the remaining bytes that are currently buffered in the deccoder to the next handler in the pipeline.
-    case forwardBytes
-    /// Discard all the remaining bytes that are currently buffered in the decoder.
-    case dropBytes
-}
-
-/// A `ChannelInboundHandler` that parses HTTP/1-style messages, converting them from
-/// unstructured bytes to a sequence of HTTP messages.
-///
-/// The `HTTPDecoder` is a generic channel handler which can produce messages in
-/// either the form of `HTTPClientResponsePart` or `HTTPServerRequestPart`: that is,
-/// it produces messages that correspond to the semantic units of HTTP produced by
-/// the remote peer.
-public class HTTPDecoder<HTTPMessageT>: ChannelInboundHandler, AnyHTTPDecoder {
-    public typealias InboundIn = ByteBuffer
-    public typealias InboundOut = HTTPMessageT
-
-    private let leftOverBytesStrategy: RemoveAfterUpgradeStrategy
-    private var parser = http_parser()
-    private var settings = http_parser_settings()
-
-    fileprivate var state = HTTPParserState()
-
-    fileprivate init(type: HTTPMessageT.Type, leftOverBytesStrategy: RemoveAfterUpgradeStrategy) {
-        /* this is a private init, the public versions only allow HTTPClientResponsePart and HTTPServerRequestPart */
-        assert(HTTPMessageT.self == HTTPClientResponsePart.self || HTTPMessageT.self == HTTPServerRequestPart.self)
-        self.leftOverBytesStrategy = leftOverBytesStrategy
     }
 
     deinit {
-        // Remove the stored reference to ChannelHandlerContext
-        self.parser.data = UnsafeMutableRawPointer(bitPattern: 0xdeadbee)
-
-        // Remove references to callbacks.
-        self.settings = http_parser_settings()
+        self.parser.deinitialize(count: 1)
+        self.parser.deallocate()
+        Unmanaged<BetterHTTPParser>.fromOpaque(UnsafeRawPointer(self.parser.pointee.data)).release()
```

Comment:
This is part of why I'm worried about the heap-allocation: this is a very straightforward use-after-free.